Bug with calling arrays.index_of_first

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 b79b4bf.b383cce
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: E:\Projects\advent2024\day5
vexe: C:\Users\drake\AppData\Local\Programs\V\V04\v.exe
vexe mtime: 2024-12-03 02:00:11

vroot: OK, value: C:\Users\drake\AppData\Local\Programs\V\V04
VMODULES: OK, value: C:\Users\drake\.vmodules
VTMP: OK, value: C:\Users\drake\AppData\Local\Temp\v_0

Git version: git version 2.41.0.windows.3
Git vroot status: Error: fatal: not a git repository (or any of the parent directories): .git
.git/config present: false

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc: N/A
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg bug.v`

```v
import arrays

fn main() {
	mut rules := [][2]int{}
	rules << [1, 2]!
	rules << [2, 3]!

	common_rule := [1, 2]!
	idx := arrays.index_of_first(rules, fn [common_rule] (idx int, rule [2]int) bool { return rule == common_rule })
	println(idx)
}
```

**What did you expect to see?**

0

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: C:/Users/drake/AppData/Local/Temp/v_0/bug.01JECGYJK0Y3Z1CJPCJFTMRVCP.tmp.c:7405: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/drake/AppData/Local/Temp/v_0/bug.01JECGYJK0Y3Z1CJPCJFTMRVCP.tmp.c:14141: error: character array initializer must be a literal, optionally enclosed in braces
... (the original output was 9 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzUyMzEzZWQyZWY0Y2JjYzQ2ZjMzZGIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.-Gfc9CxkoMN-NIBuwyUJl0gsFfHG1ObTycYJtjZTboE">Huly&reg;: <b>V_0.6-21514</b></a></sub>

--- a/vlib/v/gen/c/for.v
+++ b/vlib/v/gen/c/for.v
@@ -267,6 +267,7 @@ fn (mut g Gen) for_in_stmt(node_ ast.ForInStmt) {
 				g.writeln('\t${styp} ${c_name(node.val_var)};')
 				g.writeln('\tmemcpy(*(${styp}*)${c_name(node.val_var)}, (byte*)${right}, sizeof(${styp}));')
 			} else {
+				is_array_fixed := g.table.final_sym(node.val_type).kind == .array_fixed
 				// If val is mutable (pointer behind the scenes), we need to generate
 				// `int* val = ((int*)arr.data) + i;`
 				// instead of
@@ -276,10 +277,17 @@ fn (mut g Gen) for_in_stmt(node_ ast.ForInStmt) {
 					'((${styp}*)${opt_expr}${op_field}data)[${i}]'
 				} else if node.val_is_mut || node.val_is_ref {
 					'((${styp})${cond_var}${op_field}data) + ${i}'
+				} else if val_sym.kind == .array_fixed {
+					'((${styp}*)${cond_var}${op_field}data)[${i}]'
 				} else {
 					'((${styp}*)${cond_var}${op_field}data)[${i}]'
 				}
-				g.writeln('\t${styp} ${c_name(node.val_var)} = ${right};')
+				if !is_array_fixed {
+					g.writeln('\t${styp} ${c_name(node.val_var)} = ${right};')
+				} else {
+					g.writeln('\t${styp} ${c_name(node.val_var)} = {0};')
+					g.writeln('\tmemcpy(${c_name(node.val_var)}, ${right}, sizeof(${styp}));')
+				}
 			}
 		}
 	} else if node.kind == .array_fixed {

--- a/vlib/v/gen/c/for.v
+++ b/vlib/v/gen/c/for.v
@@ -267,7 +267,8 @@ fn (mut g Gen) for_in_stmt(node_ ast.ForInStmt) {
 				g.writeln('\t${styp} ${c_name(node.val_var)};')
 				g.writeln('\tmemcpy(*(${styp}*)${c_name(node.val_var)}, (byte*)${right}, sizeof(${styp}));')
 			} else {
-				is_array_fixed := g.table.final_sym(node.val_type).kind == .array_fixed
+				needs_memcpy := !node.val_type.is_ptr()
+					&& g.table.final_sym(node.val_type).kind == .array_fixed
 				// If val is mutable (pointer behind the scenes), we need to generate
 				// `int* val = ((int*)arr.data) + i;`
 				// instead of
@@ -282,7 +283,7 @@ fn (mut g Gen) for_in_stmt(node_ ast.ForInStmt) {
 				} else {
 					'((${styp}*)${cond_var}${op_field}data)[${i}]'
 				}
-				if !is_array_fixed {
+				if !needs_memcpy {
 					g.writeln('\t${styp} ${c_name(node.val_var)} = ${right};')
 				} else {
 					g.writeln('\t${styp} ${c_name(node.val_var)} = {0};')

--- a/vlib/v/tests/arrays_closure_fixed_array_test.v
+++ b/vlib/v/tests/arrays_closure_fixed_array_test.v
@@ -1,0 +1,17 @@
+import arrays
+
+fn test_main() {
+	mut rules := [][2]int{}
+	rules << [1, 2]!
+	rules << [2, 3]!
+
+	common_rule := [1, 2]!
+	assert arrays.index_of_first(rules, fn [common_rule] (idx int, rule [2]int) bool {
+		return rule == common_rule
+	}) == 0
+
+	common_rule2 := [2, 3]!
+	assert arrays.index_of_first(rules, fn [common_rule2] (idx int, rule [2]int) bool {
+		return rule == common_rule2
+	}) == 1
+}

